Assemble technical overview of FX

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50291 Assemble technical overview of FX**

Differential Revision: [D25908444](https://our.internmc.facebook.com/intern/diff/D25908444)

Commit:
Assemble technical overview of FX

[ghstack-poisoned]

jamesr66a commented on torch/fx/OVERVIEW.md:
````diff
@@ -0,0 +1,121 @@
+torch.testing.assert_allclose(symbolic_traced(input), module(input))
+```
+
+Here, we set up a simple Module that exercises different language features: fetching a parameter, applying an arithmetic operator, applying a submodule (linear), and applying a Tensor method. `symbolic_trace` returns an instance of GraphModule, which is in itself a subclass of `nn.Module`. We can see that the `symbolic_traced` instance runs and returns the same result as the original module insta ...
````
To make things flow better, probably should add a transition sentence mentioning something like "The following sections will walk us through the components that transform from original Module to FX IR and finally to generated Python code and a GraphModule instance"

jamesr66a commented on torch/fx/OVERVIEW.md:
```diff
@@ -0,0 +1,121 @@
+
+## Proxy ##
+
+Proxy objects are Node wrappers used by the Tracer to record operations seen during symbolic tracing. If you're doing graph transforms, you can wrap your own Proxy method around a raw Node so that you can use the overloaded operators to add additional things to a Graph.
```
I think the `If you're doing graph transforms...` part makes sense at the end after we've described what Proxies are. It would set it up like "1) here's what these are, 2) here's what you can do with them"

I think it might also make sense to have a small example of using Proxies to, e.g. append additional nodes to a Graph

ansley replied:
Is this enough of a small example that it should be in torch/fx/examples?

jamesr66a commented on torch/fx/OVERVIEW.md:
```diff
@@ -0,0 +1,121 @@
+
+Proxy objects are Node wrappers used by the Tracer to record operations seen during symbolic tracing. If you're doing graph transforms, you can wrap your own Proxy method around a raw Node so that you can use the overloaded operators to add additional things to a Graph.
+
+The mechanism through which Proxy objects record computation is [`__torch_function__`](https://pytorch.org/docs/stable/notes/extending.html). If any custom Python type defines a method named `__torch_function__`, PyTorch will invoke that `__torch_function__` implementation when an instance of that custom type is passed to a function in the `torch` namespace. In the Proxy implementation of `__torc ...
```
For the `__torch_function__` link I recommend using the anchor link (https://pytorch.org/docs/stable/notes/extending.html#extending-torch) to go directly to the `__torch_function__` section

jamesr66a replied:
In the prose, I think the causality is sort-of reversed. I'd describe the process as "1) operations on Proxy are dispatched to the `__torch_function__` handler 2) The `__torch_function__` handler records the operation in the `Graph` as a `Node` 3) The `Node` that was recorded in the `Graph` is wrapped in a `Proxy`, facilitating further application of ops on that value"

jamesr66a commented on torch/fx/OVERVIEW.md:
````diff
@@ -0,0 +1,121 @@
+  traced = symbolic_trace(m)
+```
+
+First, the parameter `x` is transformed into a Proxy object and the corresponding Node (a Node with op = “placeholder” and target = “x”) is added to the Graph. The next operation is `torch.relu`, which takes `x` as a parameter. Because `x` is a Proxy, `__torch_function__` is called on `torch.relu`. This, in turn, transforms `torch.relu` into a Proxy object and adds a `torch.relu` Node to the Grap ...
````
Specify who's transforming `x` into a proxy (the `symbolic_trace` routine). Then, the Module is run with Proxies as inputs, and recording happens via the `__torch_function__` dispatch path (which you've described)

jamesr66a commented on torch/fx/OVERVIEW.md:
```diff
@@ -0,0 +1,121 @@
+
+Here, we set up a simple Module that exercises different language features: fetching a parameter, applying an arithmetic operator, applying a submodule (linear), and applying a Tensor method. `symbolic_trace` returns an instance of GraphModule, which is in itself a subclass of `nn.Module`. We can see that the `symbolic_traced` instance runs and returns the same result as the original module insta ...
+
+## Tracer ##
```
For each of these sections (`Tracer`, `Proxy`, IR (`Graph`), it's probably a good idea to add anchor links to the respective sections in the API docs (https://pytorch.org/docs/master/fx.html) for further reference

jamesr66a commented on torch/fx/OVERVIEW.md:
```diff
@@ -0,0 +1,121 @@
+- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, *including the self argument*
+- `output` contains the output of the traced function in its `args[0]` attribute. This corresponds to the "return" statement in the Graph printout.
+
+Each Node also has a reference to the Nodes that it takes as input (`input_nodes`) and the Nodes that use it (`users`). Although Nodes are represented as a doubly-linked list, the use-def relationships form an acyclic graph and can be traversed as such.
```
Maybe something like "To facilitate easier analysis of data dependencies, `Node`s have read-only properties `input_nodes` and `users`, which specify which `Node`s in the graph are used by this node and use this node, respectively"

Also maybe we want to make `users` a read-only property :)

jamesr66a reviewed (commented):
Cool! It's going to be great to have this information publicly-accessible

Commit:
Update on "Assemble technical overview of FX"




[ghstack-poisoned]

jamesr66a commented on torch/fx/OVERVIEW.md:
````diff
@@ -0,0 +1,132 @@
+    linear_1 = self.linear(add_1)
+    clamp_1 = linear_1.clamp(min = 0.0, max = 1.0)
+    return clamp_1
+```
````
Couple things:

1) This isn't the latest code -- in particular it's missing the None assignments we added later. This can be fixed just by running the code again on master
2) The example at line 90 may confuse the reader, as it seems that this may be the code from that example rather than the one at 55

jamesr66a reviewed (commented):
Cool! There's some stuff duplicated form the `fx/__init__.py` docstring, like the description of the IR nodes. IMO we should deduplicate things and refer to that docstring as the single place this is described. This doc should be more focused on "below the cut" implementation details that are not in the public docstrings

Commit:
Update on "Assemble technical overview of FX"




[ghstack-poisoned]

jamesr66a reviewed (approved):
![image](https://user-images.githubusercontent.com/4685384/104531192-79c67e00-55c2-11eb-83b4-4e4972e3d4af.png)

Commit:
Update on "Assemble technical overview of FX"


Differential Revision: [D25908444](https://our.internmc.facebook.com/intern/diff/D25908444)

[ghstack-poisoned]
